Add the spmp extension

Fly0307 opened the pull request:
### Introduction

This PR adds the SPMP (S-level Physical Memory Protection) feature extension to the sail-riscv repository. 

###  Current Status

SPMP version: v1.0.0-rc5 is adopted, with the corresponding release link: https://github.com/riscv/riscv-spmp/releases/tag/v1.0.0-rc5
Review status: This version has passed the ARC review.

Reference for integration into the ISA manual: 
The corresponding PR link for integration into the riscv-isa-manual: https://github.com/riscv/riscv-isa-manual/pull/2521

Fly0307 committed:
Wait for another check

Fly0307 committed:
For spmp: recheck the permission settings and matching logic

Fly0307 committed:
feat(spmp): Fixed bit order errors in the implementation of SPMP RWX permission encoding

The order of RWX permission bits has been corrected from the incorrect XWR (bit [2:0]) to the correct RWX (bit [2:0]=R,W,X) to comply with the RISC-V specification. At the same time, the permission check logic and shared area rules have been updated to ensure that permission matching and access control are correctly implemented.

jordancarlin commented on config/config.json.in:
```diff
@@ -42,6 +42,14 @@
+      "tor_supported": true,
+      "na4_supported": true,
+      "napot_supported": true
+    },
```
I see why you put this in the memory section, but I'd be inclined to put this under the `extensions` key since Spmp is pretty clearly an extension.

Fly0307 committed:
refactor(spmp): Remove private modifier from enum and function declarations for better accessibility

Fly0307 committed:
Refactor SPMP Control and Register Code for Clarity and Consistency

- Removed unnecessary whitespace and improved formatting in spmp_control.sail and spmp_regs.sail.

ybc-alkaid commented on model/spmp/spmp_switch_test.sail:
@nadime15, 
Will do. Should we update the unit tests in this PR? Is model/unit_tests the preferred directory for these tests?

jordancarlin replied:
`model/unit_tests` is the directory for tests written in Sail while `test/first_party` is the directory for C or assembly tests. Unless there is a benefit to writing the tests in Sail, we prefer C/assembly tests in `test/first_party`. Even better is to get the tests upstreamed to `riscv-tests` which we will then pull in.

Fly0307 committed:
feat(spmp): Implement SPMP support and configuration in the architecture model

Fly0307 committed:
Merge branch 'master' into spmp-new

Signed-off-by: Unlearned Man <92023558+Fly0307@users.noreply.github.com>

Fly0307 committed:
feat(spmp): Add comprehensive SPMP test program and update related files

Fly0307 committed:
Merge branch 'master' into spmp-new

Signed-off-by: Unlearned Man <92023558+Fly0307@users.noreply.github.com>

Fly0307 committed:
refactor(spmp): Simplify SPMP access checks
- Delete unused MSECCFG spmp field
- Remove unused privilege parameter from SPMP access check functions

Fly0307 committed:
fix(spmp): Update memory access type references in SPMP functions

ybc-alkaid committed:
feat(spmp): rename spmpswitch to spmpen, remove old 0x1F0 CSR

- Rename internal registers spmpswitch/spmpswitchh to spmpen/spmpenh
- Delete obsolete 0x1F0/0x1F1 CSR handlers (spec uses 0x183/0x193)
- Update comments, function names, and test code accordingly

nadime15 commented:
@ybc-alkaid I would like to review this PR. If you have time, could you rebase it?

ybc-alkaid committed:
Merge upstream/master into spmp-new

ybc-alkaid commented:
@nadime15 Done, thanks for reviewing the PR!

jordancarlin reviewed (changes requested):
I haven't reviewed most of the code yet, but one early high-level comment is that this needs to be refactored to have separate config options for the different extensions instead of one global `spmp.enabled` option.

jordancarlin commented:
> @ybc-alkaid I would like to review this PR. If you have time, could you rebase it?

I think before getting too far into reviewing this PR, we need got get Sscsrind merged (#956).

ybc-alkaid committed:
refactor(spmp): split spmp.enabled into Sspmp and Sspmpen extension configs
